lib/services/role.go

```diff
@@ -1805,6 +1806,114 @@ func (set RoleSet) SessionRecordingMode(service constants.SessionRecordingServic
 	return constants.SessionRecordingModeBestEffort
 }
 
+func contains[S ~[]E, E any](s S, f func(E) (bool, error)) (bool, error) {
+	for i := range s {
+		match, err := f(s[i])
+		if err != nil {
+			return false, err
+		}
+		if match {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// matchSPIFFESVIDConditions compares a slice of SPIFFE Role Conditions against
+// a requested SPIFFE SVID generation. All fields within a condition must match,
+// but any condition in the slice can match for the function to return true.
+func matchSPIFFESVIDConditions(
+	conds []*types.SPIFFERoleCondition,
+	spiffeIDPath string,
+	dnsSANs []string,
+	ipSANs []net.IP,
+) (bool, error) {
+	return contains(conds, func(cond *types.SPIFFERoleCondition) (bool, error) {
+		// Match SPIFFE ID path.
+		expr, err := utils.CompileExpression(cond.Path)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		if !expr.MatchString(spiffeIDPath) {
+			// No match - skip to next condition.
+			return false, nil
+		}
+
+		// All DNS SANs requested must match one of the DNS SAN matchers in the
+		// condition.
+		for _, dnsSAN := range dnsSANs {
+			match, err := contains(cond.DNSSANs, func(s string) (bool, error) {
+				expr, err := utils.CompileExpression(s)
+				if err != nil {
+					return false, trace.Wrap(err)
+				}
+				return expr.MatchString(dnsSAN), nil
```

Comment:
```suggestion
				match, err := utils.MatchString(dnsSAN,s)
				return match, trace.Wrap(err)
				
```

this one caches the compiled expression

Reply:
Nice catch - I'll double check and implement

Reply:
[88bdd7d](https://github.com/gravitational/teleport/pull/38174/commits/88bdd7d5297e1363041624881046acea3a9f4e32)